File: packages/core/src/select.ts
/**
 * Select behavior machinery.
 * Contract: docs/contracts/components/select.md, "Behavior Machine".
 *
 * Pure option-list helpers plus the open/query/highlight/value transition.
 * Async loading, snippets, placement, portals, and DOM/native focus stay
 * adapter-owned.
 */

export interface SelectOptionLike {
  value: string;
  label: string;
  disabled?: boolean;
  isDisabled?: boolean;
}

export interface SelectGroupLike<T extends SelectOptionLike = SelectOptionLike> {
  label: string;
  options: T[];
}

export interface SelectOptionState {
  value: string;
  label: string;
  disabled: boolean;
}

export interface SelectContext {
  value: string;
  open: boolean;
  query: string;
  highlightedValue: string | null;
  options: SelectOptionState[];
  clearValue: string;
  searchable: boolean;
  freeform: boolean;
  disabled: boolean;
}

export type SelectEvent =
  | { type: "OPEN" }
  | { type: "CLOSE" }
  | { type: "TOGGLE" }
  | { type: "QUERY"; query: string }
  | { type: "HIGHLIGHT"; value: string }
  | { type: "HIGHLIGHT_PREV" }
  | { type: "HIGHLIGHT_NEXT" }
  | { type: "HIGHLIGHT_FIRST" }
  | { type: "HIGHLIGHT_LAST" }
  | { type: "COMMIT_HIGHLIGHTED" }
  | { type: "COMMIT_OPTION"; value: string }
  | { type: "COMMIT_FREEFORM" }
  | { type: "CLEAR" };

export type SelectEffect =
  | { type: "openChanged"; open: boolean }
  | { type: "queryChanged"; query: string }
  | { type: "valueChanged"; value: string };

export interface SelectResult {
  context: SelectContext;
  effects: SelectEffect[];
}

export function isSelectOptionDisabled(option: SelectOptionLike): boolean {
  return option.disabled === true || option.isDisabled === true;
}

/** Grouped or flat items → flat option list. */
export function flattenSelectOptions<T extends SelectOptionLike>(
  items: readonly (T | SelectGroupLike<T>)[],
): T[] {
  const first = items[0];

  if (first === undefined) {
    return [];
  }

  if ("options" in first) {
    return (items as SelectGroupLike<T>[]).flatMap((group) => group.options);
  }

  return [...(items as T[])];
}

/** Enabled options whose label contains the query (case-insensitive); empty query keeps all enabled. */
export function filterSelectOptions<T extends SelectOptionLike>(options: readonly T[], query: string): T[] {
  const enabled = options.filter((option) => !isSelectOptionDisabled(option));

  if (query.length === 0) {
    return enabled;
  }

  const lowered = query.toLowerCase();

  return enabled.filter((option) => option.label.toLowerCase().includes(lowered));
}

/** Per-group label filter; groups left empty by the filter are dropped. */
export function filterSelectGroups<T extends SelectOptionLike>(
  groups: readonly SelectGroupLike<T>[],
  query: string,
): SelectGroupLike<T>[] {
  const lowered = query.toLowerCase();

  return groups
    .map((group) => ({
      ...group,
      options: group.options.filter((option) => option.label.toLowerCase().includes(lowered)),
    }))
    .filter((group) => group.options.length > 0);
}

/** Highlight index on open: the selected option when present, else the first. */
export function selectOpenHighlightIndex<T extends SelectOptionLike>(
  filtered: readonly T[],
  selectedValue: string | null,
): number {
  if (selectedValue === null) {
    return 0;
  }

  const index = filtered.findIndex((option) => option.value === selectedValue);

  return index < 0 ? 0 : index;
}

export function selectFreeformEnabled(context: SelectContext): boolean {
  return context.searchable && context.freeform;
}

export function selectMatchesQuery(label: string, query: string): boolean {
  if (query.length === 0) {
    return true;
  }

  return label.toLowerCase().includes(query.toLowerCase());
}

/** Visible options for the current query, including disabled rows. */
export function selectVisibleOptions(context: SelectContext): SelectOptionState[] {
  return context.options.filter((option) => selectMatchesQuery(option.label, context.query));
}

export function selectEnabledVisibleValues(context: SelectContext): string[] {
  return selectVisibleOptions(context)
    .filter((option) => !option.disabled)
    .map((option) => option.value);
}

export function selectCommittedQuery(context: SelectContext): string {
  if (context.value === context.clearValue) {
    return "";
  }

  return context.options.find((option) => option.value === context.value)?.label ?? "";
}

/** Opening highlight: selected enabled visible option, else first enabled visible, else null. */
export function selectOpenHighlightValue(context: SelectContext): string | null {
  const enabled = selectEnabledVisibleValues(context);

  if (enabled.length === 0) {
    return null;
  }

  if (context.value !== context.clearValue && enabled.includes(context.value)) {
    return context.value;
  }

  return enabled[0] ?? null;
}

function inert(context: SelectContext): SelectResult {
  return { context, effects: [] };
}

function pushOpen(effects: SelectEffect[], previous: boolean, next: boolean): void {
  if (previous !== next) {
    effects.push({ type: "openChanged", open: next });
  }
}

function pushQuery(effects: SelectEffect[], previous: string, next: string): void {
  if (previous !== next) {
    effects.push({ type: "queryChanged", query: next });
  }
}

function pushValue(effects: SelectEffect[], previous: string, next: string): void {
  if (previous !== next) {
    effects.push({ type: "valueChanged", value: next });
  }
}

function orderedEffects(
  previous: SelectContext,
  next: Pick<SelectContext, "open" | "query" | "value">,
  includeQuery: boolean,
): SelectEffect[] {
  const effects: SelectEffect[] = [];

  pushOpen(effects, previous.open, next.open);

  if (includeQuery) {
    pushQuery(effects, previous.query, next.query);
  }

  pushValue(effects, previous.value, next.value);

  return effects;
}

function openList(context: SelectContext): SelectResult {
  if (context.open) {
    return inert(context);
  }

  const next: SelectContext = {
    ...context,
    open: true,
    highlightedValue: selectOpenHighlightValue({ ...context, open: true }),
  };

  return { context: next, effects: orderedEffects(context, next, false) };
}

function closeList(context: SelectContext): SelectResult {
  if (!context.open) {
    return inert(context);
  }

  const next: SelectContext = {
    ...context,
    open: false,
    query: selectFreeformEnabled(context) ? context.query : selectCommittedQuery(context),
  };

  return { context: next, effects: orderedEffects(context, next, false) };
}

function findOption(context: SelectContext, value: string): SelectOptionState | undefined {
  return context.options.find((option) => option.value === value);
}

function moveHighlight(context: SelectContext, direction: 1 | -1): SelectResult {
  if (!context.open) {
    return openList(context);
  }

  const enabled = selectEnabledVisibleValues(context);

  if (enabled.length === 0) {
    if (context.highlightedValue === null) {
      return inert(context);
    }

    return { context: { ...context, highlightedValue: null }, effects: [] };
  }

  const currentIndex = context.highlightedValue === null ? -1 : enabled.indexOf(context.highlightedValue);
  let nextIndex: number;

  if (currentIndex < 0) {
    nextIndex = direction === 1 ? 0 : enabled.length - 1;
  } else {
    nextIndex = Math.max(0, Math.min(enabled.length - 1, currentIndex + direction));
  }

  const highlightedValue = enabled[nextIndex] ?? null;

  if (highlightedValue === context.highlightedValue) {
    return inert(context);
  }

  return { context: { ...context, highlightedValue }, effects: [] };
}

function jumpHighlight(context: SelectContext, position: "first" | "last"): SelectResult {
  if (!context.open) {
    return inert(context);
  }

  const enabled = selectEnabledVisibleValues(context);
  const highlightedValue = position === "first" ? (enabled[0] ?? null) : (enabled[enabled.length - 1] ?? null);

  if (highlightedValue === context.highlightedValue) {
    return inert(context);
  }

  return { context: { ...context, highlightedValue }, effects: [] };
}

function commitOption(context: SelectContext, value: string): SelectResult {
  const option = findOption(context, value);

  if (!option || option.disabled) {
    return inert(context);
  }

  const next: SelectContext = {
    ...context,
    value: option.value,
    query: option.label,
    open: false,
    highlightedValue: option.value,
  };

  return { context: next, effects: orderedEffects(context, next, true) };
}

function commitFreeform(context: SelectContext): SelectResult {
  if (!selectFreeformEnabled(context) || context.highlightedValue !== null || context.query === context.value) {
    return inert(context);
  }

  const next: SelectContext = {
    ...context,
    value: context.query,
    open: false,
    highlightedValue: null,
  };

  return { context: next, effects: orderedEffects(context, next, false) };
}

export function selectTransition(context: SelectContext, event: SelectEvent): SelectResult {
  if (context.disabled) {
    return inert(context);
  }

  switch (event.type) {
    case "OPEN":
      return openList(context);
    case "CLOSE":
      return closeList(context);
    case "TOGGLE":
      return context.open ? closeList(context) : openList(context);
    case "QUERY": {
      const next: SelectContext = {
        ...context,
        query: event.query,
        open: true,
      };
      next.highlightedValue = selectOpenHighlightValue(next);

      return { context: next, effects: orderedEffects(context, next, true) };
    }
    case "HIGHLIGHT": {
      const option = findOption(context, event.value);

      if (!option || option.disabled || !selectMatchesQuery(option.label, context.query)) {
        return inert(context);
      }

      if (context.highlightedValue === option.value) {
        return inert(context);
      }

      return { context: { ...context, highlightedValue: option.value }, effects: [] };
    }
    case "HIGHLIGHT_PREV":
      return moveHighlight(context, -1);
    case "HIGHLIGHT_NEXT":
      return moveHighlight(context, 1);
    case "HIGHLIGHT_FIRST":
      return jumpHighlight(context, "first");
    case "HIGHLIGHT_LAST":
      return jumpHighlight(context, "last");
    case "COMMIT_HIGHLIGHTED":
      if (context.highlightedValue !== null) {
        return commitOption(context, context.highlightedValue);
      }

      return commitFreeform(context);
    case "COMMIT_OPTION":
      return commitOption(context, event.value);
    case "COMMIT_FREEFORM":
      return commitFreeform(context);
    case "CLEAR": {
      const next: SelectContext = {
        ...context,
        value: context.clearValue,
        query: "",
        highlightedValue: context.open ? selectOpenHighlightValue({ ...context, value: context.clearValue, query: "" }) : null,
      };

      return { context: next, effects: orderedEffects(context, next, true) };
    }
  }
}

File: packages/contracts/headless/src/select.rs
//! Select machine. Mirror of core `select.ts`.
//!
//! Pure open/query/highlight/value transitions. Adapters own async loading,
//! placement, portals, and DOM/native focus.

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SelectOptionState {
    pub value: String,
    pub label: String,
    pub disabled: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SelectContext {
    pub value: String,
    pub open: bool,
    pub query: String,
    pub highlighted_value: Option<String>,
    pub options: Vec<SelectOptionState>,
    pub clear_value: String,
    pub searchable: bool,
    pub freeform: bool,
    pub disabled: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SelectEvent {
    Open,
    Close,
    Toggle,
    Query { query: String },
    Highlight { value: String },
    HighlightPrev,
    HighlightNext,
    HighlightFirst,
    HighlightLast,
    CommitHighlighted,
    CommitOption { value: String },
    CommitFreeform,
    Clear,
    OptionsChanged { options: Vec<SelectOptionState> },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SelectEffect {
    OpenChanged { open: bool },
    QueryChanged { query: String },
    ValueChanged { value: String },
}

pub fn select_freeform_enabled(context: &SelectContext) -> bool {
    context.searchable && context.freeform
}

pub fn select_matches_query(label: &str, query: &str) -> bool {
    query.is_empty() || label.to_lowercase().contains(&query.to_lowercase())
}

pub fn select_visible_options(context: &SelectContext) -> Vec<&SelectOptionState> {
    context
        .options
        .iter()
        .filter(|option| select_matches_query(&option.label, &context.query))
        .collect()
}

pub fn select_enabled_visible_values(context: &SelectContext) -> Vec<String> {
    select_visible_options(context)
        .into_iter()
        .filter(|option| !option.disabled)
        .map(|option| option.value.clone())
        .collect()
}

pub fn select_committed_query(context: &SelectContext) -> String {
    if context.value == context.clear_value {
        return String::new();
    }

    context
        .options
        .iter()
        .find(|option| option.value == context.value)
        .map(|option| option.label.clone())
        .unwrap_or_default()
}

pub fn select_query_highlight_value(context: &SelectContext) -> Option<String> {
    select_enabled_visible_values(context).into_iter().next()
}

pub fn select_open_highlight_value(context: &SelectContext) -> Option<String> {
    let enabled = select_enabled_visible_values(context);

    if enabled.is_empty() {
        return None;
    }

    if context.value != context.clear_value && enabled.iter().any(|value| value == &context.value) {
        return Some(context.value.clone());
    }

    enabled.into_iter().next()
}

fn inert(context: SelectContext) -> (SelectContext, Vec<SelectEffect>) {
    (context, Vec::new())
}

fn ordered_effects(
    previous: &SelectContext,
    next: &SelectContext,
    include_query: bool,
) -> Vec<SelectEffect> {
    let mut effects = Vec::new();

    if previous.open != next.open {
        effects.push(SelectEffect::OpenChanged { open: next.open });
    }

    if include_query && previous.query != next.query {
        effects.push(SelectEffect::QueryChanged {
            query: next.query.clone(),
        });
    }

    if previous.value != next.value {
        effects.push(SelectEffect::ValueChanged {
            value: next.value.clone(),
        });
    }

    effects
}

fn open_list(context: SelectContext) -> (SelectContext, Vec<SelectEffect>) {
    if context.open {
        return inert(context);
    }

    let mut next = context.clone();
    next.open = true;
    next.highlighted_value = select_open_highlight_value(&next);
    let effects = ordered_effects(&context, &next, false);
    (next, effects)
}

fn close_list(context: SelectContext) -> (SelectContext, Vec<SelectEffect>) {
    if !context.open {
        return inert(context);
    }

    let mut next = context.clone();
    next.open = false;
    if !select_freeform_enabled(&context) {
        next.query = select_committed_query(&context);
    }
    let effects = ordered_effects(&context, &next, false);
    (next, effects)
}

fn find_option<'a>(context: &'a SelectContext, value: &str) -> Option<&'a SelectOptionState> {
    context.options.iter().find(|option| option.value == value)
}

fn move_highlight(context: SelectContext, direction: i8) -> (SelectContext, Vec<SelectEffect>) {
    if !context.open {
        return open_list(context);
    }

    let enabled = select_enabled_visible_values(&context);

    if enabled.is_empty() {
        if context.highlighted_value.is_none() {
            return inert(context);
        }

        let mut next = context;
        next.highlighted_value = None;
        return (next, Vec::new());
    }

    let current_index = context
        .highlighted_value
        .as_ref()
        .and_then(|value| enabled.iter().position(|item| item == value));
    let last = enabled.len() - 1;
    let next_index = match current_index {
        None if direction > 0 => 0,
        None => last,
        Some(index) if direction > 0 => index.saturating_add(1).min(last),
        Some(index) => index.saturating_sub(1),
    };
    let highlighted_value = enabled.get(next_index).cloned();

    if highlighted_value == context.highlighted_value {
        return inert(context);
    }

    let mut next = context;
    next.highlighted_value = highlighted_value;
    (next, Vec::new())
}

fn jump_highlight(context: SelectContext, first: bool) -> (SelectContext, Vec<SelectEffect>) {
    if !context.open {
        return inert(context);
    }

    let enabled = select_enabled_visible_values(&context);
    let highlighted_value = if first {
        enabled.first().cloned()
    } else {
        enabled.last().cloned()
    };

    if highlighted_value == context.highlighted_value {
        return inert(context);
    }

    let mut next = context;
    next.highlighted_value = highlighted_value;
    (next, Vec::new())
}

fn commit_option(context: SelectContext, value: &str) -> (SelectContext, Vec<SelectEffect>) {
    let Some(option) = find_option(&context, value) else {
        return inert(context);
    };

    if option.disabled {
        return inert(context);
    }

    let mut next = context.clone();
    next.value = option.value.clone();
    next.query = option.label.clone();
    next.open = false;
    next.highlighted_value = Some(option.value.clone());
    let effects = ordered_effects(&context, &next, true);
    (next, effects)
}

fn commit_freeform(context: SelectContext) -> (SelectContext, Vec<SelectEffect>) {
    if !select_freeform_enabled(&context)
        || context.highlighted_value.is_some()
        || context.query == context.value
    {
        return inert(context);
    }

    let mut next = context.clone();
    next.value = context.query.clone();
    next.open = false;
    next.highlighted_value = None;
    let effects = ordered_effects(&context, &next, false);
    (next, effects)
}

pub fn select_transition(
    context: SelectContext,
    event: SelectEvent,
) -> (SelectContext, Vec<SelectEffect>) {
    if context.disabled {
        return inert(context);
    }

    match event {
        SelectEvent::Open => open_list(context),
        SelectEvent::Close => close_list(context),
        SelectEvent::Toggle => {
            if context.open {
                close_list(context)
            } else {
                open_list(context)
            }
        }
        SelectEvent::Query { query } => {
            let mut next = context.clone();
            next.query = query;
            next.open = true;
            next.highlighted_value = select_query_highlight_value(&next);
            let effects = ordered_effects(&context, &next, true);
            (next, effects)
        }
        SelectEvent::Highlight { value } => {
            let highlighted = {
                let Some(option) = find_option(&context, &value) else {
                    return inert(context);
                };

                if option.disabled || !select_matches_query(&option.label, &context.query) {
                    return inert(context);
                }

                if context.highlighted_value.as_deref() == Some(option.value.as_str()) {
                    return inert(context);
                }

                option.value.clone()
            };
            let mut next = context;
            next.highlighted_value = Some(highlighted);
            (next, Vec::new())
        }
        SelectEvent::HighlightPrev => move_highlight(context, -1),
        SelectEvent::HighlightNext => move_highlight(context, 1),
        SelectEvent::HighlightFirst => jump_highlight(context, true),
        SelectEvent::HighlightLast => jump_highlight(context, false),
        SelectEvent::CommitHighlighted => match context.highlighted_value.clone() {
            Some(value) => commit_option(context, &value),
            None => commit_freeform(context),
        },
        SelectEvent::CommitOption { value } => commit_option(context, &value),
        SelectEvent::CommitFreeform => commit_freeform(context),
        SelectEvent::Clear => {
            let mut next = context.clone();
            next.value = context.clear_value.clone();
            next.query = String::new();
            next.highlighted_value = if context.open {
                select_open_highlight_value(&next)
            } else {
                None
            };
            let effects = ordered_effects(&context, &next, true);
            (next, effects)
        }
        SelectEvent::OptionsChanged { options } => {
            let mut next = context;
            next.options = options;
            if next.open {
                next.highlighted_value = select_query_highlight_value(&next);
            }
            (next, Vec::new())
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn fruit() -> Vec<SelectOptionState> {
        vec![
            SelectOptionState {
                value: "apple".into(),
                label: "Apple".into(),
                disabled: false,
            },
            SelectOptionState {
                value: "banana".into(),
                label: "Banana".into(),
                disabled: false,
            },
            SelectOptionState {
                value: "cherry".into(),
                label: "Cherry".into(),
                disabled: false,
            },
            SelectOptionState {
                value: "spinach".into(),
                label: "Spinach".into(),
                disabled: true,
            },
        ]
    }

    fn context() -> SelectContext {
        SelectContext {
            value: String::new(),
            open: false,
            query: String::new(),
            highlighted_value: None,
            options: fruit(),
            clear_value: "apple".into(),
            searchable: false,
            freeform: false,
            disabled: false,
        }
    }

    #[test]
    fn disabled_context_is_inert_in_every_direction() {
        let mut disabled = context();
        disabled.disabled = true;
        for event in [
            SelectEvent::Open,
            SelectEvent::Toggle,
            SelectEvent::HighlightNext,
            SelectEvent::CommitOption {
                value: "banana".into(),
            },
            SelectEvent::Clear,
        ] {
            let (next, effects) = select_transition(disabled.clone(), event);
            assert_eq!(next, disabled);
            assert!(effects.is_empty());
        }
    }

    #[test]
    fn disabled_option_commit_is_inert() {
        let mut open = context();
        open.open = true;
        let (next, effects) = select_transition(
            open.clone(),
            SelectEvent::CommitOption {
                value: "spinach".into(),
            },
        );
        assert_eq!(next, open);
        assert!(effects.is_empty());
    }

    #[test]
    fn highlight_navigation_skips_disabled_and_lands_on_enabled_bounds() {
        let mut open = context();
        open.open = true;
        open.highlighted_value = Some("cherry".into());
        let (next, effects) = select_transition(open, SelectEvent::HighlightNext);
        assert_eq!(next.highlighted_value.as_deref(), Some("cherry"));
        assert!(effects.is_empty());

        let mut open = context();
        open.open = true;
        let (next, _) = select_transition(open, SelectEvent::HighlightLast);
        assert_eq!(next.highlighted_value.as_deref(), Some("cherry"));
    }

    #[test]
    fn options_changed_revalidates_a_stale_highlight() {
        let mut open = context();
        open.open = true;
        open.highlighted_value = Some("banana".into());
        let mut options = fruit();
        options.retain(|option| option.value != "banana");
        let (next, effects) = select_transition(open, SelectEvent::OptionsChanged { options });
        assert_eq!(next.highlighted_value.as_deref(), Some("apple"));
        assert!(effects.is_empty());
        let (committed, commit_effects) = select_transition(next, SelectEvent::CommitHighlighted);
        assert_eq!(committed.value, "apple");
        assert!(commit_effects
            .iter()
            .any(|effect| matches!(effect, SelectEffect::ValueChanged { value } if value == "apple")));
        assert!(commit_effects.iter().all(
            |effect| !matches!(effect, SelectEffect::ValueChanged { value } if value == "banana")
        ));
    }

    #[test]
    fn clear_restores_the_authored_default() {
        let mut selected = context();
        selected.value = "banana".into();
        let (next, effects) = select_transition(selected, SelectEvent::Clear);
        assert_eq!(next.value, "apple");
        assert_eq!(
            effects,
            vec![SelectEffect::ValueChanged {
                value: "apple".into()
            }]
        );
    }

    #[test]
    fn effects_are_ordered_open_then_query_then_value() {
        let mut searchable = context();
        searchable.searchable = true;
        searchable.freeform = true;
        searchable.open = true;
        searchable.query = "mango".into();
        searchable.highlighted_value = None;
        let (next, effects) = select_transition(searchable, SelectEvent::CommitFreeform);
        assert_eq!(next.value, "mango");
        assert!(!next.open);
        assert_eq!(
            effects,
            vec![
                SelectEffect::OpenChanged { open: false },
                SelectEffect::ValueChanged {
                    value: "mango".into()
                },
            ]
        );
    }

    #[test]
    fn query_opens_and_highlights_the_first_enabled_match() {
        let mut searchable = context();
        searchable.searchable = true;
        let (next, effects) = select_transition(
            searchable,
            SelectEvent::Query {
                query: "ban".into(),
            },
        );
        assert!(next.open);
        assert_eq!(next.query, "ban");
        assert_eq!(next.highlighted_value.as_deref(), Some("banana"));
        assert_eq!(
            effects,
            vec![
                SelectEffect::OpenChanged { open: true },
                SelectEffect::QueryChanged {
                    query: "ban".into()
                },
            ]
        );
    }
}
